docs/building-modules/module-interfaces.md

```diff
@@ -1,9 +1,282 @@
+		SuggestionsMinimumDistance: 2,
+		RunE:                       client.ValidateCmd,
+	}
+	txCmd.AddCommand(
+		GetMultiSignCommand(cdc),
+		GetSignCommand(cdc),
+	)
+	return txCmd
+}
+```
+
+### Query Commands
+
+[Queries](./messages-and-queries.md#queries) allow users to gather information about the application or network state; they are routed by the application and processed by the module in which they are defined. Query commands typically have their own `query.go` file in the module `x/moduleName/client/cli` folder. Like transaction commands, they are specified in getter functions and have the prefix `GetCmdQuery`. Here is an example of a query command from the [nameservice module CLI](https://cosmos.network/docs/tutorial/cli.html#queries):
+
+```go
+func GetCmdWhois(queryRoute string, cdc *codec.Codec) *cobra.Command {
+ return &cobra.Command{
+   Use: "whois [name]",
+   Short: "Query whois info of name",
+   Args: cobra.ExactArgs(1),
+   RunE: func(cmd *cobra.Command, args []string) error {
+     cliCtx := context.NewCLIContext().WithCodec(cdc)
+     name := args[0]
+
+     res, err := cliCtx.QueryWithData(fmt.Sprintf("custom/%s/whois/%s", queryRoute, name), nil)
+     if err != nil {
+       fmt.Printf("could not resolve whois - %s \n", string(name))
+       return nil
+     }
+
+     var out nameservice.Whois
+     cdc.MustUnmarshalJSON(res, &out)
+     return cliCtx.PrintOutput(out)
+   },
+ }
+}
+
+```
+
+This query returns the address that owns a particular name. The getter function does the following:
+
+- **`codec` and `queryRoute`.** In addition to taking in the application `codec`, query command getters also take a `queryRoute` used to construct a path [Baseapp](../core/baseapp.md#query-routing) uses to route the query in the application.
+- **Construct the command.** Read the [Cobra Documentation](https://github.com/spf13/cobra) and the [transaction command](#transaction-commands) example above for more information. The user must type `whois` and provide the `name` they are querying for as the only argument.
+- **`RunE`.** The function should be specified as a `RunE` to allow for errors to be returned. This function encapsulates all of the logic to create a new query that is ready to be relayed to nodes.
+  + The function should first initialize a new [`CLIContext`](./query-lifecycle.md#clicontext) with the application `codec`.
+  + If applicable, the `CLIContext` is used to retrieve any parameters (e.g. the query originator's address to be used in the query) and marshal them with the query parameter type, in preparation to be relayed to a node. There are no `CLIContext` parameters in this case because the query does not involve any information about the user.
+  + The `queryRoute` is used to construct a route Baseapp will use to route the query to the appropriate [querier](./querier.md). Module queries are `custom` type queries (some SDK modules have exceptions, such as `auth` and `gov` module queries).
```

Comment:
This bullet point should be about:

- The `QueryWithData` method (a short explainer)
- The standard way to format a query `path` -> "queryCategory/queryRoute/queryType/arg1/arg2/...", where queryCategory can be app/p2p/custom/store (explain each briefly), queryRoute is usually module's name, queryType is the actual specific module-query, and then the args. 

Actually maybe this will be hard to fit in a bullet point, but should defnitily be specified in this section.

Also when mentioning baseapp, link to it plz!

Reply:
@gzhao408 if you have time could you please address this point? It's all that's needed to merge.

If not please tell me, we'll have to close this PR and pick up your work in a new branch

Reply:
@gamarin2 addressed

Reply:
Thank you!!